File: 240817/아름다운 수/beautiful-number.py
n = int(input())
ans = 0
seq = list()

def is_beautiful():
    idx = 0
    while idx < n:
        if idx + seq[idx] - 1 >= n:
            return False

        for i in range(idx, idx + seq[idx]):
            if (seq[i] != seq[idx]):
                return False
        
        idx += seq[idx]

    return True

def calc(cnt):
    global ans;

    if cnt == n:
        if is_beautiful():
            ans += 1

        return

    for i in range(1, 5):
        seq.append(i)
        calc(cnt + 1)
        seq.pop()

calc(0)
print(ans)

File: 240817/아름다운 수/beautiful-number.java
import java.util.Scanner;
import java.util.ArrayList;

public class Main {
    static int ans = 0;
    static int n;
    static ArrayList<Integer> nums = new ArrayList<>();

    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        n = s.nextInt();

        countBtfNum(0);
        System.out.println(ans);
        
    }

    public static boolean isBeautiful(){
        int idx = 0;
        
        while (idx < n) {
            int targetNum = nums.get(idx);

            if (idx + targetNum - 1 >= n){
                return false;
            }

            for (int i=idx; i<idx + targetNum ; i++){
                if (targetNum != nums.get(i)){
                    return false;
                }
            }

            idx += targetNum;
        }
        return true;
    }

    public static void countBtfNum(int num){
        if (num == n) {
            if (isBeautiful())
                ans++;
            return;
        }

        for (int i=1; i<5; i++){
            nums.add(i);
            countBtfNum(num + 1);
            nums.remove(nums.size()-1);
        }
    }
}